library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -2088,8 +2108,14 @@ impl<T, A: Allocator> VecDeque<T, A> {
     #[inline]
     #[stable(feature = "append", since = "1.4.0")]
     pub fn append(&mut self, other: &mut Self) {
-        // naive impl
-        self.extend(other.drain(..));
+        self.reserve(other.len());
+        unsafe {
+            let (left, right) = other.as_slices();
+            self.append_slice(left);
+            self.append_slice(right);
+        }
+        // Silently drop values in `other`.
+        other.tail = other.head;
```

Comment:
This doesn't seem "panic safe", as in, if append_slice were to panic, then the elements would exist in both VecDeques afterwards, causing double frees and other corruptions. And indeed you even added an explicit panic (assert) in append_slice, so it can indeed panic.

Reply:
You're right, I haven't thought of that.
But I wonder if it's always possible to write panic-safe code. Here I could `other.wrap_add(other.tail, some_len)` after each `append_slice` to make sure values are dropped from `other` only after they are copied and `self.head` is updated. But that relies on `wrap_add` no panicking, which sadly not the case.

Reply:
you could append both slices in a single function, then you can do all necessary calculations that might panic up-front before doing any assignments. That way the structs will either be in their old or in their new state.

Reply:
Comment addressed~ Now pointers are updated after copying.